File: Mobiles/react-native/ios/QuickPizza/QuickPizzaDebugModule.m
#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_MODULE(QuickPizzaDebug, NSObject)

RCT_EXTERN_METHOD(blockMainThread:(nonnull NSNumber *)durationMs
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

@end

File: Mobiles/react-native/ios/QuickPizza/QuickPizzaDebugModule.swift
import Darwin
import Foundation
import React

@objc(QuickPizzaDebug)
class QuickPizzaDebug: NSObject {
  @objc
  static func requiresMainQueueSetup() -> Bool {
    false
  }

  /// Blocks the main run loop for demo frozen-frame injection (Faro CADisplayLink).
  @objc(blockMainThread:resolver:rejecter:)
  func blockMainThread(
    _ durationMs: NSNumber,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    let clampedMs = min(max(durationMs.doubleValue, 1), 10_000)
    DispatchQueue.main.async {
      let deadline = Date().addingTimeInterval(clampedMs / 1000.0)
      while Date() < deadline {
        sched_yield()
      }
      resolve(nil)
    }
  }
}
